content/blog/kallisti5/2018-05-19_rune_-_haiku_images_on_arm.md

```diff
@@ -0,0 +1,65 @@
+
+Fedora's arm-image-installer tool:
+
+  * Takes the Fedora ARM image you download
+  * You specify the target board (Like the Raspberry Pi 3)
+  * Takes the properiety bootloader artifacts from your local system
+  * Selects the correct FDT file
+  * Burns all of this to an SD card ready to boot on your device
+
+So, end users use the "arm-image-installer" tool to "burn" the ARM image to an SD card (and doing a bunch of magic under the covers)
+
+{{< alert-warning "Spoiler" "Haiku on ARM is in its infancy. Don't expect any magic. 'rpi2' will likely work the 'best' which involves loading out kernel and spitting out information to the serial uart.">}}
+
+# Rune Image
+
+With arm-image-installer in my mind, I went to work on "Rune Image". Rune is a tool written in Rust which:
+
+  * Takes a generic ["Haiku ARM" image](https://download.haiku-os.org/nightly-images/arm/) you provide (hrev51955 or later)
+  * Reads a JSON document in one of our [git repositories](https://github.com/haiku/firmware) via HTTP
+  * Grabs the needed prioperitery bits (as well as u-boot binaries + documented source locations) via HTTP
+  * Selects the correct FDT dtb file
+  * Burns all of this to an SD card (or modifies an existing image) ready to boot on your device.
+
+As long as our ARM build architecture (currently ARMv7) is compatible with the hardware, generic images can run on any ARM hardware as long as we have driver support, a valid Device Tree binary for the board, and the requirements defined in our [repository manifest](https://github.com/haiku/firmware/blob/master/manifest.json).
+
+A quick (poorly edited) [demo video](https://vimeo.com/271101255).
+
+You can find a pre-compiled binary for Linux x86_64 [here](https://github.com/kallisti5/rune-image/releases/tag/v0.1.0).
+
+# Why Rust?
+
+I've had a few people ask why I wrote rune in Rust when Haiku is 99% C++. I actually have a lot of good reasons...
+
+  * Cross-platform.
+    * Rune compiles on Linux, OS X, Windows (untested), and Haiku (our Rust port needs work though).
+    * This tool is intended to be run by end users. It needs to be portable to their operating systems.
+  * Low-level.
+    * Rune doesn't need mtools or other external requirements.
+    * Rune writes to FAT partitions in our images _directly_.
+  * Minimal requirements.
+    * Easy HTTP, JSON, etc, with built-in libraries.
+
+# Future
+
+Lets avoid the sensational "Haiku now runs on ARM" news.
+Haiku has run on ARM for quite some time, these changes simply enable us to increase our development velocity while simplifying the distribution of Haiku ARM images to a single "generic ARMv7" image.
```

Comment:
Yay, now I need to install Rust and compile a tool not integrated in the buildsystem just so I can test if it boots.
Do you know the most popular request (ok, that was just geist) is "give me an easy way to run the kernel in qemu in a single command, so I don't have to mess with buildsystems and obscure tools and I can play with the kernel debugger"

Reply:
Actually, no. I provide pre-compiled self-contained binaries per the blog article. :-)
Just Linux-x86_64 for now... but soon I should auto-generated OS X and windows binaries as well.